### ISSUE TITLE
Literal information lost after change of type of class attribute

### ISSUE THREAD

**Issue by lijunkang12** [microsoft/pyright#1940]
**To Reproduce**
```python
from typing import Literal
Bar = Literal[0]

class Foo():
    def __init__(self):
        self.tuple_bar: tuple[Bar, ...] = (0, 0, 0)
        self.list_bar: list[Bar] = list(self.tuple_bar)
    
    def list_of_bar(self) -> list[Bar]:
        list_bar: list[Bar] = list(self.tuple_bar)
        return list_bar
```
In the previous example, Pyright throws an error
```
Expression of type "list[int]" cannot be assigned to declared type "list[Bar]"
  TypeVar "_T@list" is invariant
    "int" cannot be assigned to type "Bar"
```
complaining about the `list_of_bar` method. However, it does not complain about the attribute `self.list_bar` in `__init__`.

**VS Code extension or command-line**
Pylance 2021.6.0 (with Pyright 1.1.146)

**Comment by lijunkang12** [comment]
Another relevant issue. Following the previous code, If I write
```Python
for index in self.tuple_bar:
    pass
```
index will be typed as `int`, instead of `Bar`. This will cause typing error when I want to write something such as
```Python
my_dict: dict[Bar, str] = {0: '0'}
bla = [my_dict[index] for index in self.tuple_bar]
```
since in the definition of `bla`, `index` has type `int` instead of `Bar`. How can I get around this problem?

**Comment by lijunkang12** [comment]
Another issue with Literal:
```Python
foo: tuple[Bar, ...] = (0, 0, 0) # no error
foo2: tuple[Bar, ...] = tuple(0 for _ in range(3)) # no error
foo3: tuple[Bar, ...] = (0, ) * 3 # error
foo4: tuple[Bar, ...] = (0, ) + (0, ) + (0, ) # error
```
The last two expressions get the following error message:
```
Expression of type "Tuple[int, ...]" cannot be assigned to declared type "tuple[Bar, ...]"
  Tuple entry 1 is incorrect type
    "int" cannot be assigned to type "Bar"
```

**Comment by erictraut** [comment]
The first two cases are related and are due to a bug. I've introduced a fix, and it will be included in the next release.

The third case involving arithmetic operations on tuples is expected behavior given pyright's inference rules for tuples and the fact that bidirectional type inference cannot be applied to function calls (like `__mul__` or `__add__`). You would need to work around this using something like the following:

```python
foo1: tuple[Bar, ...] = (0,)
foo3 = foo1 * 3
```

**Comment by lijunkang12** [comment]
Thanks for your answer and for introducing a fix! Glad that I found a real bug this time.

By the way, I would like to ask a question not relevant to this issue. Since I am using Pyright via Pylance, where should I report bugs of type checking? Here or rather in the repo pylance-release?

**Comment by erictraut** [comment]
You can report them in either the pyright or pylance-release repos. If it's a type checking bug that is reported to pylance-release, I typically create a corresponding tracking bug in pyright.

**Comment by erictraut** [comment]
This is included in pyright 1.1.147, which I just published. It will also be included in the next release of pylance.
